squidpy/gr/_sepal.py

```diff
@@ -0,0 +1,341 @@
+from functools import singledispatch
+from typing_extensions import Literal  # < 3.8
+
+from scanpy import logging as logg
+from anndata import AnnData
+from scanpy.get import _get_obs_rep
+
+from numba import njit
+from scipy.sparse import spmatrix
+from sklearn.metrics import pairwise_distances
+import numpy as np
+import pandas as pd
+
+from squidpy._docs import d, inject_docs
+from squidpy._utils import Signal, SigQueue, parallelize, _get_n_cores
+from squidpy.gr._utils import (
+    _save_data,
+    _assert_spatial_basis,
+    _assert_connectivity_key,
+    _assert_non_empty_sequence,
+)
+from squidpy._constants._pkg_constants import Key
+
+__all__ = ["sepal"]
+
+
+@d.dedent
+@inject_docs(key=Key.obsp.spatial_conn())
+def sepal(
+    adata: AnnData,
+    max_nbrs: Literal[4, 6],
+    genes: Optional[Union[str, Sequence[str]]] = None,
+    n_iter: Optional[int] = 30000,
+    dt: float = 0.001,
+    thres: float = 1e-8,
+    connectivity_key: str = Key.obsp.spatial_conn(),
+    spatial_key: str = Key.obsm.spatial,
+    layer: Optional[str] = None,
+    use_raw: bool = False,
+    copy: bool = False,
+    n_jobs: Optional[int] = None,
+    backend: str = "loky",
+    show_progress_bar: bool = True,
+) -> Optional[pd.DataFrame]:
+    """
+    Identify spatially variable genes with Sepal.
+
+    Sepal is a method that simulate a diffusion process to quantify spatial structure in tissue.
+    See  :cite:`andersson2021` for reference.
+
+    Parameters
+    ----------
+    %(adata)s
+    max_nbrs
+        Maximum number of neighbors of a node in the graph, either:
+
+            - 4 for a square-grid (ST, Dbit-seq).
+            - 6 for a hexagonal-grid (Visium).
+    genes
+        List of gene names, as stored in :attr:`anndata.AnnData.var_names`, used to compute sepal score.
+
+        If `None`, it's computed :attr:`anndata.AnnData.var` ``['highly_variable']``, if present.
+        Otherwise,it's computed for all genes.
+    n_iter
+        Maximum number of iterations for the diffusion simulation.
+    dt
+        Time step added in diffusion simulation.
+    thres
+        Entropy threshold for convergence of diffusion simulation.
+    %(conn_key)s
+    %(spatial_key)s
+    layer
+        Layer in :attr:`anndata.AnnData.layers` to use. If `None`, use :attr:`anndata.AnnData.X`.
+    use_raw
+        Whether to access :attr:`anndata.AnnData.raw`.
+    %(copy)s
+    %(parallelize)s
+
+    Returns
+    -------
+    If ``copy = True``, returns a :class:`pandas.DataFrame` with the sepal scores.
+
+    Otherwise, adds to ``adata.uns`` the following key:
+
+        - :attr:`anndata.AnnData.uns` ``['sepal_score']`` - sepal score.
+
+    Notes
+    -----
+    If some genes in :attr:`anndata.AnnData.uns` ``['sepal_score']`` are `NaN`,
+    consider re-running the function with increased `n_iter`.
+    """
+    _assert_connectivity_key(adata, connectivity_key)
+    _assert_spatial_basis(adata, key=spatial_key)
+
```

Comment:
Please add check for the maximum neighbors if it's in `(4, 6)` and raise ValueError if it isn't.
I know it's being checked later down the code, but not after some operations with non-trivial cost.

Reply:
it's now added before any computation (after checking it graph is sparse)